### ISSUE TITLE
[FEA] Refactor expensive string APIs, combining two kernel calls into one

### ISSUE THREAD

**Issue by ttnghia** [rapidsai/cudf#16073]
Currently, most string APIs perform their work by calling their main kernel twice, one for computing the output string sizes and one for writing the output strings.

There is a technique that has a potential performance improvement:
 * Firstly, allocate an output buffer that is large enough to write the output without knowing the output string sizes.
 * The main kernel is then called only once, writing both the output string sizes and the output strings into the buffer above.
 * Finally, a string factory function is called to gather the strings and sizes above.

Note that the aforementioned technique can improve the performance of expensive string APIs, when the second kernel call is more expensive than the final string gather step. Otherwise it does not provide any benefit.

**Comment by davidwendt** [comment]
In general, the gather step would have to be faster than the size computation which seems unlikely since the gather itself is 2 kernel calls -- one computes the offsets and the 2nd gathers the memory. Also, the temporary memory could be significantly higher (and effect performance) since each output row needed for the gather requires a pointer/size pair (~16 bytes/row). Not to mention the over-allocated temporary buffer.

So I would not expect this to be a performance improvement in general for strings processing but we could do some benchmarking/profiling on a case-by-case basis as appropriate.

**Comment by ttnghia** [comment]
Right, this may not work in all cases but only case-by-case basis. The more expensive kernels, the more speedup we can get (up to 2X). I'm not sure which string APIs can be improved the most, maybe regex like?

**Comment by davidwendt** [comment]
> Right, this may not work in all cases but only case-by-case basis. The more expensive kernels, the more speedup we can get (up to 2X). I'm not sure which string APIs can be improved the most, maybe regex like?

Again, I would doubt the size calculation would be faster than building a index vector and then calling gather.
This would really only apply to creating brand new strings (not copying, not extracting, etc) and _usually_ these only create small strings. For copy, extract, and even replace for example for long strings, the existing code already converts these operations into a gather (since the strings components exist in memory but just need specific ordering) which is faster only because the memcpy is faster (not the size calculation as proposed here). This may be a technique you can use as well to avoid the over-allocated temporary buffer.

So I appreciate the suggestion and I think the spirit of what you are proposing is actually already in place especially for supporting large strings. I don't think there is any additional work here to do.
